speechbrain/lobes/models/discrete/SQ-Codec.py

```diff
@@ -0,0 +1,1497 @@
+        os.makedirs(save_path)
+
+    # Download the file with progress bar
+    zip_filename = hf_hub_download(
+        repo_id=repo_id, filename=filename, cache_dir=save_path
+    )
+
+    # Extract the file
+    with zipfile.ZipFile(zip_filename, "r") as zip_ref:
+        zip_ref.extractall(save_path)
+
+    # Remove the downloaded ZIP file
+    os.remove(zip_filename)
+    print(f"File downloaded, extracted to '{save_path}', and ZIP file removed.")
+
+
+def decimal_to_ternary_matrix(decimals, D):
+    """
+    Convert a tensor of decimal numbers to a D*T ternary matrix for each batch.
+
+    Parameters
+    ----------
+    decimals : torch.Tensor
+        A 2D tensor of decimal numbers with shape (B, T), where B is the batch size
+        and T is the number of elements in each batch.
+    D : int
+        Number of ternary digits to represent each number (depth).
+
+    Returns
+    -------
+    torch.Tensor
+        A 3D tensor of shape (B, D, T) where each slice along the first dimension
+        corresponds to a batch, and each column is represented as a ternary number.
+    """
+    B, T = decimals.shape
+    ternary_matrix = torch.zeros((B, D, T), dtype=torch.long)
+    for pos in range(D):
+        ternary_matrix[:, pos, :] = decimals % 3  # Modulo operation
+        decimals //= 3  # Floor division for next ternary digit
+
+    return ternary_matrix
+
+
+def ternary_matrix_to_decimal(matrix):
+    """
+    Convert a B*D*N ternary matrix to a 2D array of decimal numbers for each batch.
+
+    Parameters
+    ----------
+    matrix : numpy.ndarray
+        A 3D numpy array of shape (B, D, N), where B is the batch size, D is the number
+        of ternary digits, and N is the number of ternary numbers in each batch.
+
+    Returns
+    -------
+    numpy.ndarray
+        A 2D numpy array of shape (B, N), where each value represents the decimal
+        equivalent of the corresponding ternary number in the input matrix.
+    """
+    B, D, N = (
+        matrix.shape
+    )  # B is the batch size, D is the number of digits, N is the number of ternary numbers
+    powers_of_three = 3 ** np.arange(D)  # [3^0, 3^1, ..., 3^(D-1)]
+
+    # Reshape powers_of_three for broadcasting: [D] -> [1, D, 1]
+    powers_of_three = powers_of_three[:, np.newaxis]  # Shape [D, 1]
+
+    # Compute dot product using broadcasting: matrix * powers_of_three along D axis
+    decimals = np.sum(matrix * powers_of_three, axis=1)  # Sum along the D axis
+
+    return decimals
+
+
+def get_padding(kernel_size, dilation=1):
+    """
+    Computes the padding size for a given kernel size and dilation.
+
+    Arguments
+    ---------
+    kernel_size : int
+        Size of the convolutional kernel.
+    dilation : int, optional
+        Dilation factor for convolution (default is 1).
+
+    Returns
+    -------
+    int
+        Calculated padding size.
+    """
+    return int((kernel_size * dilation - dilation) / 2)
+
+
+class round_func5(InplaceFunction):
```

Comment:
Couldn't you just use "torch.round(5 * input) / 5" in the code when necessary, this is building an overcomplex torch custom operation for something that could be automatically inferred from torch.

Reply:
If these function ends up being mandatory, unittests will certainly be necessary to check the gradients and shape etc.

Reply:
When you use torch.round(5 * input) / 5 directly, the gradient for torch.round is not defined (or zero). PyTorch treats it as a non-differentiable operation, and gradients will not propagate through it which is essential in scalar quantization.

Reply:
i created a "CustomRoundingFunction" that handle all these rounding and backprop